### ISSUE TITLE
`prepare deployment` step is very slow

### ISSUE THREAD

**Issue by beyhan** [cloudfoundry/bosh#1847]
**Symptoms**
The `prepare deployment` step during a deployment is very slow. 

**Analysis**
When some deployments are with cloud-config and some are without, then the time to gather and parse all the needed data (manifest, parsing it and doing some network checks) increases. This depends on the number of deployments and network configs which are managed by the BOSH director.

**Mitigations**
Don't mix deployments with cloud-config and without.

**Identified Issues**
Deployment times increase significantly.

**Comment by cppforlife** [comment]
can you provide more information about timing? which parts do you think are
taking long time? is it trying to iterate over bunch of ips in a big range
(tries to do that in this type of heterogeneous env)?

**Comment by beyhan** [comment]
yes, it is iterating over all non cloud config manifests to get the reserved ips and compute the reserved ranges. Here is an example log for how long it takes (I have removed the reserved ip ranges):
```
D, [2017-12-01T12:05:22 #22912] [task:1487595] DEBUG -- DirectorJobRunner: reserved ranges 
D, [2017-12-01T12:05:22 #22912] [task:1487595] DEBUG -- DirectorJobRunner: (0.038069s) SELECT * FROM "deployments" WHERE ((("deployments"."id" NOT IN (SELECT "deployments_configs"."deployment_id" FROM "configs" INNER JOIN "deployments_configs" ON ("deployments_configs"."config_id" = "configs"."id") WHER
I, [2017-12-01T12:21:03 #22912] [task:1487595]  INFO -- DirectorJobRunner: Following networks and individual IPs are reserved by non-cloud-config deployments: ...
D, [2017-12-01T12:21:03 #22912] [task:1487595] DEBUG -- DirectorJobRunner: reserved ranges ...
```

**Comment by cppforlife** [comment]
chatted with @voelzmo today. possible slowness due to https://github.com/cloudfoundry/bosh/blob/6e743b8db05bf9656a6686eb84764da923559ad7/src/bosh-director/lib/bosh/director/deployment_plan/global_network_resolver.rb#L48. let's comment out this method (since it's purely for debugging) and see what are the times.

**Comment by KaiHofstetter** [comment]
We saw multiple `SELECT * FROM "instances"` in  the task trace log during `prepare deployment`, which takes between 0.5 - 1 seconds. The reason for these expensive SQL queries is caused by the creation of the `LocalDnsEncoder`, which is described in more detail in [this tracker comment](https://www.pivotaltracker.com/story/show/153991487/comments/186188289). 

We are thinking about reusing the `LocalDnsEncoder` for all instance_plans of a deployment, which would increase the performance of bosh deployment with a large number of instances (>100) by a large factor.

Any thoughts about this?

**Comment by dpb587-pivotal** [comment]
I think it would be optimal for there to be a single `LocalDnsEncoder` which is shared across the deploy process. It might be a little messy trying to pass it through everywhere, but from your investigations it sounds like it would be worth the performance improvements.

Are there any concerns about threading/concurrency and the singleton behaviors?

**Comment by mfine30** [comment]
Seems related to: https://www.pivotaltracker.com/story/show/159469692
